text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,460 @@
+
+Note that the server does not need to know in what mode the client operates. The server just needs to ensure that it sends a response to every request. The Synchronized or Pipelined mode is enabled purely by client behavior.
+
+### Server Response
+
+The Server replies to every `Export` request received from the client by a corresponding `Export` response. The ID field of the response is set equal to the ID field of the request.
+
+#### Result Code
+
+`Export` response includes a `result_code` field, which indicates whether the server was able to successfully process the received data. Possible values for `result_code` field are:
+
+- `Success` - telemetry data is successfully processed by the server.
+
+- `FailedNoneRetryable` - processing of telemetry data failed. The client MUST NOT retry sending the same telemetry data. The telemetry data MUST be dropped. This for example can happen when the request contains bad data and cannot be deserialized or otherwise processed by the server. The client SHOULD maintain a counter of such dropped data.
+
+- `FailedRetryable` - processing of telemetry data failed. The client SHOULD record the error and may retry exporting the same data. This can happen when the server is temporarily unable to process the data.
+
+#### Throttling
+
+OTLP allows backpressure signalling.
+
+If the server is unable to keep up with the pace of data it receives from the client then it SHOULD signal that fact to the client. The client MUST then throttle itself to avoid overwhelming the server.
+
+`Export` response includes a `throttle_period_millisec` field, which indicates the minimum duration in milliseconds that the client MUST wait before sending the next `Export` request. The waiting period starts from the moment the client receives such `Export` response. The value of 0 for `throttle_period_millisec` indicates that the client may send the next `Export` request immediately and is an indication that no throttling is requested by the server.
+
+This field is used by the server to signal backpressure to the client. The value of `throttle_period_millisec` is determined by the server and is implementation dependant. The server SHOULD choose a `throttle_period_millisec` value that is big enough to give the server time to recover, yet is not too big to cause the client to drop data while it is throttled.
+
+### gRPC Transport
+
+#### Service Definition
+
+`Hello` request and response is done via a unary gRPC call.
+
+`Export` requests and responses are delivered using bidirectional gRPC streams.
+
+This is OTLP over gRPC Service definition:
+
+```
+service StreamExporter {
+  // Hello handshake.
+  rpc Hello(HelloRequest) returns (HelloResponse) {}
+
+  // Sends a batch of telemetry data.
+  rpc Export(stream ExportRequest) returns (stream ExportResponse) {}
```

Comment:
>One problem with stream which I did not think about is setting a deadline per message. We do want to have a deadline probably for every request correct? I am not sure if this is important or not but please also think about.

Yes, client implementations definitely will need some sort of timeout after which they should consider the message to be lost if they don't hear from the server. At that point they have a choice: declare the message dropped (and increment some internal "dropped" counter) or retry sending the message. I think we should go with "dropped" approach, because retrying is risky, we don't know why the server is unable to respond to this message and can end up sending the same message over and over again and potentially overload the server.

I will add this to the spec.

> Also if we do go with unary RPCs we can remove the ack id because then it is very easy to map request with the response, also the piepline will be implemented using async unary rpc.

See my other response regarding unary performance and why I think we should use streaming instead.

> I am leaning more towards unary rpcs because you also removed the optimization to not send the node/resource.

I did not remove the node/resource optimization. I changes the Export message format to allow to include multiple nodes.

This is OpenCensus export format:
```
message ExportTraceServiceRequest {
  Node node = 1;
  repeated Span spans = 2;
  Resource resource = 3;
}
```
This allows one Node, one Resource and multiple Spans.

I suggest to change this to following:
```
message ExportRequest {
  uint64 id = 1;
  repeated NodeSpans nodeSpans = 2;
  repeated NodeMetrics nodeMetrics = 3;
}
message NodeSpans {
  Node node = 1;
  Resource resource = 2;
  repeated Span spans = 3;
}
```
`NodeSpans` looks like `ExportTraceServiceRequest` in OpenCensus. In OLTP you can have many of these in one `ExportRequest` (via `repeated NodeSpans nodeSpans = 2` field). This is important for Collector use-case.

Collector receives telemetry data from different sources. Different sources means different `Node` infromation. One of the jobs of a Collector is to combine telemetry data into batches for more efficient processing. However because of how OpenCensus `ExportTraceServiceRequest` schema is defined it is impossible to batch telemetry data from different sources (different Nodes) into one export request. This can have dramatic performance impact on the efficiency because it can increase the number of Export requests that Collector must do by orders of magnitude.

We observed exactly this problem with OpenCensus protocol just last week.

My proposal fixes this problem, one `ExportRequest` can contain telemetry data from many Nodes.

Reply:
The optimization that you removed is that Node/Resource can be sent once per stream. You did keep the optimization to send Node/Resource once for multiple data

Reply:
I think we can still have the same optimization per stream. It will be a bit more difficult for intermediary nodes like otelsvc to handle this optimization but I think it is doable.

For data recieved from telemetry sources (applications) if Resource is not specified in ExportRequest then the last Resource is applied ("last" means anything that is listed in the same ExportRequest in earlier batch or in the previous ExportRequest). This will allow telemetry sources which always use the same Resource to specify a single SpanBatch in ExportRequest and specify Resource only once, in the first request. Typically ExportRequest will contain only one element in spanBatches array.

otelsvc needs to be a bit more complex to handle this optimization. otelsvc needs to combine SpanBatch-es received from different streams into one Export request (and create bigger batches). In that case ExportRequest will contain an array of SpanBatch-es. If the SpanBatch received from a particular stream does not have a Resource specified than otelsvc will add the Resource field before combining it into multi-batch request. To be able to do so otelsvc receiver must cache the last received Resource per each stream. The otelsvc batcher will also be able to do batch compacting, it can eliminate unnecessary repetitions of the same Resource inside multiple batches received from the same source when creating a multi-batch request, like this:

![image](https://user-images.githubusercontent.com/4194920/63362217-7da78f80-c33f-11e9-8f9d-33fcc210f7ad.png)